File: src/NodeParser/IdentifierNodeParser.ts
import ts from "typescript";
import type { Context, NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { UnknownNodeError } from "../Error/Errors.js";

/**
 * Resolves identifiers whose value is a compile-time constant or a class-like declaration.
 */
export class IdentifierNodeParser implements SubNodeParser {
    constructor(
        private readonly childNodeParser: NodeParser,
        private readonly checker: ts.TypeChecker,
    ) {}

    supportsNode(node: ts.Identifier): boolean {
        return node.kind === ts.SyntaxKind.Identifier;
    }

    createType(node: ts.Identifier, context: Context): BaseType {
        const symbol = this.checker.getSymbolAtLocation(node);
        if (!symbol) {
            throw new UnknownNodeError(node);
        }

        if (symbol.flags & ts.SymbolFlags.Alias) {
            return this.createTypeFromDeclaration(
                this.checker.getAliasedSymbol(symbol).valueDeclaration,
                context,
                node,
            );
        }

        return this.createTypeFromDeclaration(symbol.valueDeclaration, context, node);
    }

    protected createTypeFromDeclaration(
        decl: ts.Declaration | undefined,
        context: Context,
        node: ts.Identifier,
    ): BaseType {
        if (
            decl &&
            ts.isVariableDeclaration(decl) &&
            decl.initializer &&
            ts.getCombinedNodeFlags(decl) & ts.NodeFlags.Const
        ) {
            return this.childNodeParser.createType(decl.initializer, context);
        }

        if (
            decl &&
            (ts.isClassDeclaration(decl) ||
                ts.isInterfaceDeclaration(decl) ||
                ts.isTypeAliasDeclaration(decl))
        ) {
            return this.childNodeParser.createType(decl, context);
        }

        if (decl && ts.isParameter(decl)) {
            const parameterType = this.getConstructorParameterArgumentType(decl, context);
            if (parameterType) {
                return parameterType;
            }
        }

        throw new UnknownNodeError(node);
    }

    protected getConstructorParameterArgumentType(
        parameter: ts.ParameterDeclaration,
        context: Context,
    ): BaseType | undefined {
        const args = context.getArguments();
        if (!args.length) {
            return undefined;
        }

        const parent = parameter.parent;
        if (!ts.isFunctionLike(parent)) {
            return undefined;
        }

        const index = parent.parameters.indexOf(parameter);
        if (index < 0 || index >= args.length) {
            return undefined;
        }

        return args[index];
    }
}

File: internal/parser/identifier_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// IdentifierNodeParser resolves identifiers whose value is a compile-time
// constant (src/NodeParser/IdentifierNodeParser.ts). Only `const` variables
// are resolved.
type IdentifierNodeParser struct {
	childNodeParser NodeParser
	checker         *checker.Checker
}

func NewIdentifierNodeParser(childNodeParser NodeParser, typeChecker *checker.Checker) *IdentifierNodeParser {
	return &IdentifierNodeParser{childNodeParser: childNodeParser, checker: typeChecker}
}

func (p *IdentifierNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindIdentifier
}

func (p *IdentifierNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	symbol := p.checker.GetSymbolAtLocation(node)
	if symbol == nil {
		panic(NewUnknownNodeError(node))
	}

	declaration := symbol.ValueDeclaration
	if declaration != nil && ast.IsVariableDeclaration(declaration) &&
		declaration.Initializer() != nil &&
		ast.GetCombinedNodeFlags(declaration)&ast.NodeFlagsConst != 0 {
		return p.childNodeParser.CreateType(declaration.Initializer(), ctx, nil)
	}

	panic(NewUnknownNodeError(node))
}
